src/main/java/net/oauth/jsontoken/AsyncJsonTokenParser.java

```diff
@@ -0,0 +1,148 @@
+   *   based on a signature algorithm, the signer, and key ids.
+   * @param executor an executor to run the tasks before and after getting the verifiers
+   * @param checkers an array of checkers that validates the parameters in the JSON token.
+   */
+  public AsyncJsonTokenParser(
+      Clock clock, AsyncVerifierProviders asyncVerifierProviders, Executor executor, Checker... checkers) {
+    super(clock, checkers);
+    this.asyncVerifierProviders = Preconditions.checkNotNull(asyncVerifierProviders);
+    this.executor = Preconditions.checkNotNull(executor);
+  }
+
+  /**
+   * Verifies that the jsonToken has a valid signature and valid standard claims
+   * (iat, exp). Uses {@link AsyncVerifierProviders} to obtain the secret key.
+   * This method is not expected to throw exceptions when returning a future. However,
+   * when getting the result of the future, an {@link ExecutionException} may be thrown
+   * in which {@link ExecutionException#getCause()} follows the same possible exceptions
+   * as thrown by {@link JsonTokenParser#verify(JsonToken)}.
+   *
+   * @param jsonToken
+   * @return a {@link ListenableFuture} that will return a Void
+   */
+  public ListenableFuture<Void> verify(JsonToken jsonToken) {
+    ListenableFuture<List<Verifier>> futureVerifiers = provideVerifiers(jsonToken);
+    // Use AsyncFunction instead of Function to allow for checked exceptions to propagate forward
+    AsyncFunction<List<Verifier>, Void> verifyFunction =
+        verifiers -> Futures.immediateFuture(verifyAndReturnVoid(jsonToken, verifiers));
+
+    return Futures.transformAsync(futureVerifiers, verifyFunction, executor);
+  }
+
+  /**
+   * Parses, and verifies, a JSON Token.
+   * This method is not expected to throw exceptions when returning a future. However,
+   * when getting the result of the future, an {@link ExecutionException} may be thrown
+   * in which {@link ExecutionException#getCause()} follows the same possible exceptions
+   * thrown by {@link JsonTokenParser#verifyAndDeserialize(String)}.
+   *
+   * @param tokenString the serialized token that is to parsed and verified.
+   * @return a {@link ListenableFuture} that will return the deserialized {@link JsonObject},
+   * suitable for passing to the constructor of {@link JsonToken}
+   * or equivalent constructor of {@link JsonToken} subclasses.
+   */
+  public ListenableFuture<JsonToken> verifyAndDeserialize(String tokenString) {
+    ListenableFuture<String> futureTokenString = Futures.immediateFuture(tokenString);
+    AsyncFunction<String, JsonToken> verifyFunction =
+        tokenStringInput -> {
+          JsonToken jsonToken = deserialize(tokenStringInput);
+          Function<Void, JsonToken> returnFunction = unused -> jsonToken;
+          return Futures.transform(verify(jsonToken), returnFunction, executor);
+        };
+
+    return Futures.transformAsync(futureTokenString, verifyFunction, executor);
+  }
```

Comment:
We can do the initial non-blocking part (deserialize) immediately without wrapping it into a future and transforming it. If it fails we could catch it and return Futures.immediateFailedFuture. Same with provideVerifiers

Reply:
Changed here and in provideVerifiers.